libyul/optimiser/RedundantStoreEliminator.cpp

```diff
@@ -0,0 +1,447 @@
+			Operation{
+				Location::Memory,
+				Effect::Read,
+				identifierNameIfSSA(_functionCall.arguments.at(arguments - 4)),
+				identifierNameIfSSA(_functionCall.arguments.at(arguments - 3))
+			},
+			// Unknown read includes unknown write.
+			Operation{Location::Storage, Effect::Read, {}, {}},
+			Operation{
+				Location::Memory,
+				Effect::Write,
+				identifierNameIfSSA(_functionCall.arguments.at(arguments - 2)),
+				identifierNameIfSSA(_functionCall.arguments.at(arguments - 1))
+			}
+		};
+	}
+	case Instruction::CREATE:
+	case Instruction::CREATE2:
+		return vector<Operation>{
+			Operation{
+				Location::Memory,
+				Effect::Read,
+				identifierNameIfSSA(_functionCall.arguments.at(1)),
+				identifierNameIfSSA(_functionCall.arguments.at(2))
+			},
+			// Unknown read includes unknown write.
+			Operation{Location::Storage, Effect::Read, {}, {}},
+		};
+	case Instruction::MSIZE:
+		// This is just to satisfy the assert below.
+		return vector<Operation>{};
+	default:
+		break;
+	}
+
+	yulAssert(
+		evmasm::SemanticInformation::storage(*instruction) ==
+		evmasm::SemanticInformation::None,
+		""
+	);
+	yulAssert(
+		evmasm::SemanticInformation::memory(*instruction) ==
+		evmasm::SemanticInformation::None,
+		""
+	);
+	yulAssert(
+		sideEffects.memory == SideEffects::Effect::None &&
+		sideEffects.storage == SideEffects::Effect::None,
+		""
+	);
+	return {};
+}
+
+void RedundantStoreEliminator::applyOperation(RedundantStoreEliminator::Operation const& _operation)
+{
+	for (auto& [statement, state]: m_stores[YulString{}])
+		if (state == State::Undecided)
+		{
+			Operation const& storeOperation = m_storeOperations.at(statement);
+			if (_operation.effect == Effect::Read && !knownUnrelated(storeOperation, _operation))
+				state = State::Used;
+			else if (_operation.effect == Effect::Write && knownCovered(storeOperation, _operation))
+				state = State::Unused;
+		}
+}
+
+bool RedundantStoreEliminator::knownUnrelated(
+	RedundantStoreEliminator::Operation const& _op1,
+	RedundantStoreEliminator::Operation const& _op2
+) const
+{
+	KnowledgeBase knowledge(m_dialect, m_ssaValues);
+
+	if (_op1.location != _op2.location)
+		return true;
+	if (_op1.location == Location::Storage)
+	{
+		if (_op1.start && _op2.start)
+			return knowledge.knownToBeDifferent(*_op1.start, *_op2.start);
+	}
+	else
+	{
+		u256 largestPositive = (u256(1) << 128) - 1;
+
+		yulAssert(_op1.location == Location::Memory, "");
+		if (_op1.length && knowledge.knownToBeZero(*_op1.length))
+			return true;
+		if (_op2.length && knowledge.knownToBeZero(*_op2.length))
+			return true;
+
+		// 1.start + 1.length <= 2.start
+		if (_op1.length && _op2.start && _op1.start)
+		{
+			optional<u256> length1 = knowledge.valueIfKnownConstant(*_op1.length);
+			optional<u256> diff = knowledge.differenceIfKnownConstant(*_op2.start, *_op1.start);
+			// diff = 2.start - 1.start
+			if (length1 && diff)
+				if (
+					*length1 <= *diff &&
+					*length1 <= largestPositive &&
```

Comment:
Actually isn't this redundant?

Reply:
It is :)

Similarly in `knownCovered()`, I think that `e.length <= i.length` is already covered by `i.start <= e.start` and `e.start + e.length <= i.start + i.length` together.